File: Round-1/Array/Numbers of good pairs/num_of_good_pairs.cpp
#include <iostream>
#include <vector>
using namespace std;

// Sol - 2 > optimise solution time complexity O(n)  space complexity O(n)
// optimise class
class Solution
{
public:
    int numIdenticalPairs(vector<int> &nums)
    {
        vector<int32_t> cnt(100);
        int ans = 0;
        for (auto i : nums)
        {
            ans += cnt[i - 1];
            ++cnt[i - 1];
        }
        return ans;
    }
};

// Sol - 1 > brute force approach to do this O(n^2)
// brute force class
// class Solution
// {
// public:
//     int numIdenticalPairs(vector<int> &nums)
//     {
//         int32_t pairCount = 0;
//         for (int32_t i = 0; i < nums.size() - 1; ++i)
//         {
//             for (int32_t j = i + 1; j < nums.size(); ++j)
//             {
//                 if (nums[i] == nums[j])
//                 {
//                     pairCount++;
//                 }
//             }
//         }
//         return pairCount;
//     }
// };
int main()
{
    Solution Sol;
    vector<int> arr{1, 2, 3, 1, 1, 3};
    Sol.numIdenticalPairs(arr);
    return 0;
}

File: Round-1/Array/Numbers of good pairs/num_of_good_pairs.js
// Sol - 1 > brute force approach to do this O(n^2)
// var numIdenticalPairs = function (nums) {
//     let pairCount = 0;
//     for (let i = 0; i < nums.length - 1; i++) {
//         for (let j = i + 1; j < nums.length; j++) {
//             if (nums[i] === nums[j]) {
//                 pairCount++;
//             }
//         }
//     }
//     return pairCount;
// };



// Sol - 2 > Optimise solution  - time O(n) - space O(n)
var numIdenticalPairs = function (nums) {
    var numIdenticalPairs = function (nums) {
        const good_pairs = new Array(100).fill(0);
        let ans = 0;
        nums.forEach(num => {
            ans += good_pairs[num - 1];
            ++good_pairs[num - 1];
        })
        return ans;
    };
};

console.log(numIdenticalPairs([1, 2, 3, 1, 1, 3]))